tests/test_bootstrap.py

```diff
@@ -0,0 +1,523 @@
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.ctx.bootstrap = bs
+
+        # test that prepare_build_dir runs (notice that we mock
+        # any file/dir creation so we can speed up the tests)
+        bs.prepare_build_dir()
+        # make sure that the open command has been called only once
+        mock_open.assert_called_with("project.properties", "w")
+        mock_open.reset_mock()
+
+        # check that the symlink was made 4 times and that
+        self.assertEqual(
+            len(mock_sh_ln.call_args_list), len(mock_listdir.return_value)
+        )
+        for i, directory in enumerate(mock_listdir.return_value):
+            self.assertTrue(
+                mock_sh_ln.call_args_list[i][0][1].endswith(directory)
+            )
+
+
+class GenericBootstrapTest(BaseClassSetupBootstrap):
+    """
+    An inherited class of `BaseClassSetupBootstrap` which will extends his
+    functionality by adding some generic bootstrap tests, so this way we can
+    test all our sub modules of :mod:`~pythonforandroid.bootstraps` from within
+    this module.
+
+    .. warning:: This is supposed to be used as a base class, so please, don't
+                 use this directly.
+    """
+
+    @property
+    def bootstrap_name(self):
+        """Subclasses must have property 'bootstrap_name'. It should be the
+        name of the bootstrap to test"""
+        raise NotImplementedError("Not implemented in GenericBootstrapTest")
+
+    @mock.patch("pythonforandroid.bootstraps.service_only.open", create=True)
+    @mock.patch("pythonforandroid.bootstraps.webview.open", create=True)
+    @mock.patch("pythonforandroid.bootstraps.sdl2.open", create=True)
+    @mock.patch("pythonforandroid.distribution.open", create=True)
+    @mock.patch("pythonforandroid.build.Context.get_python_install_dir")
+    @mock.patch(
+        "pythonforandroid.python.GuestPythonRecipe.create_python_bundle"
+    )
+    @mock.patch(
+        "pythonforandroid.python.GuestPythonRecipe.compile_python_files"
+    )
+    @mock.patch("pythonforandroid.bootstrap.Bootstrap.strip_libraries")
+    @mock.patch("pythonforandroid.util.exists")
+    @mock.patch("pythonforandroid.bootstrap.os.path.isfile")
+    @mock.patch("pythonforandroid.bootstrap.os.path.exists")
+    @mock.patch("pythonforandroid.bootstrap.os.unlink")
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy2")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.Command")
+    @mock.patch("pythonforandroid.bootstrap.sh.zip")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.sh.cp")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_run_distribute(
```

Comment:
I think the following mocks are not necessary as no code path would reach them in the given tests.

- mock_isdir
- mock_sh_mkdir
- mock_sh_zip
- mock_sh_command
- mock_sh_ln
- mock_shutil_copy
- mock_shutil_copy2
- mock_os_unlink
- mock_os_path_isfile
- mock_get_python_install_dir

Removing them would make the code more readable.
Maybe as a rule of thumb we should try to prove `assert mock.called is True` or even `assert mock.mock_calls != []` to show the mock is actually being called? What do you think?

Reply:
I think that you are right, we should check that all the mocks are being called, so I revised all this pull request and I think that now all the mocks are being tested in some way or other so we should be free of unnecessary mocks...so please, let me know if you find one so I can fix it before merge this